File: src/main/java/com/github/sorusclient/client/ui/framework/constraint/Flexible.java
package com.github.sorusclient.client.ui.framework.constraint;

import com.github.sorusclient.client.ui.framework.Component;
import com.github.sorusclient.client.util.Color;

import java.util.List;

public class Flexible implements Constraint {

    @Override
    public double getXValue(Component.Runtime componentRuntime) {
        return 0;
    }

    @Override
    public double getYValue(Component.Runtime componentRuntime) {
        return 0;
    }

    @Override
    public double getWidthValue(Component.Runtime componentRuntime) {
        List<double[]> placedComponents = componentRuntime.getParent().getPlacedComponents();

        double maxLeft = -Double.MAX_VALUE;
        double maxRight = Double.MAX_VALUE;

        for (double[] component : placedComponents) {
            double padding = Math.max(componentRuntime.getPadding(), component[4]);

            boolean intersectingY = componentRuntime.getY() + componentRuntime.getHeight() / 2 + padding > component[1] - component[3] / 2 + 1 && componentRuntime.getY() - componentRuntime.getHeight() / 2 + 1 < component[1] + component[3] / 2 + padding;
            boolean canInteractLeft = component[0] < componentRuntime.getX();

            if (intersectingY && canInteractLeft) {
                double componentRight = component[0] + component[2] / 2 + padding;
                maxLeft = Math.max(maxLeft, componentRight);
            }

            boolean canInteractRight = component[0] > componentRuntime.getX();

            if (intersectingY && canInteractRight) {
                double componentLeft = component[0] - component[2] / 2 - padding;
                maxRight = Math.min(maxRight, componentLeft);
            }
        }

        return Math.max(0, maxRight - maxLeft);
    }

    @Override
    public double getHeightValue(Component.Runtime componentRuntime) {
        List<double[]> placedComponents = componentRuntime.getParent().getPlacedComponents();

        double maxTop = -Double.MAX_VALUE;
        double maxBottom = Double.MAX_VALUE;

        for (double[] component : placedComponents) {
            double padding = Math.max(componentRuntime.getPadding(), component[4]);

            boolean intersectingX = componentRuntime.getX() + componentRuntime.getWidth() / 2 + padding > component[0] - component[2] / 2 + 1 && componentRuntime.getX() - componentRuntime.getWidth() / 2 + 1 < component[0] + component[2] / 2 + padding;
            boolean canInteractTop = component[1] < componentRuntime.getY();

            if (intersectingX && canInteractTop) {
                double componentBottom = component[1] + component[3] / 2 + padding;
                maxTop = Math.max(maxTop, componentBottom);
            }

            boolean canInteractBottom = component[1] > componentRuntime.getY();

            if (intersectingX && canInteractBottom) {
                double componentTop = component[1] - component[3] / 2 - padding;
                maxBottom = Math.min(maxBottom, componentTop);
            }
        }

        return Math.max(0, maxBottom - maxTop);
    }

    @Override
    public double getPaddingValue(Component.Runtime componentRuntime) {
        return 0;
    }

    @Override
    public Color getColorValue(Component.Runtime componentRuntime) {
        return null;
    }

    @Override
    public String getStringValue(Component.Runtime componentRuntime) {
        return null;
    }

    @Override
    public double getCornerRadiusValue(Component.Runtime componentRuntime) {
        return 0;
    }

}

File: src/main/kotlin/com/github/sorusclient/client/ui/framework/constraint/Flexible.kt
package com.github.sorusclient.client.ui.framework.constraint

import com.github.sorusclient.client.ui.framework.Component
import com.github.sorusclient.client.util.Color

class Flexible : Constraint {
    override fun getXValue(componentRuntime: Component.Runtime): Double {
        return 0.0
    }

    override fun getYValue(componentRuntime: Component.Runtime): Double {
        return 0.0
    }

    override fun getWidthValue(componentRuntime: Component.Runtime): Double {
        val placedComponents = componentRuntime.getParent()!!.placedComponents
        var maxLeft = -Double.MAX_VALUE
        var maxRight = Double.MAX_VALUE
        for (component in placedComponents) {
            val bottomPadding = componentRuntime.bottomPadding.coerceAtLeast(component.second[6])
            val topPadding = componentRuntime.topPadding.coerceAtLeast(component.second[7])
            val intersectingY = componentRuntime.y + componentRuntime.height / 2 + bottomPadding > component.second[1] - component.second[3] / 2 + 1 && componentRuntime.y - componentRuntime.height / 2 + 1 < component.second[1] + component.second[3] / 2 + topPadding

            val leftPadding = componentRuntime.rightPadding.coerceAtLeast(component.second[4])
            val canInteractLeft = component.second[0] < componentRuntime.x
            if (intersectingY && canInteractLeft) {
                val componentRight = component.second[0] + component.second[2] / 2 + leftPadding
                maxLeft = maxLeft.coerceAtLeast(componentRight)
            }

            val rightPadding = componentRuntime.leftPadding.coerceAtLeast(component.second[5])
            val canInteractRight = component.second[0] > componentRuntime.x
            if (intersectingY && canInteractRight) {
                val componentLeft = component.second[0] - component.second[2] / 2 - rightPadding
                maxRight = maxRight.coerceAtMost(componentLeft)
            }
        }
        return 0.0.coerceAtLeast(maxRight - maxLeft)
    }

    override fun getHeightValue(componentRuntime: Component.Runtime): Double {
        val placedComponents = componentRuntime.getParent()!!.placedComponents
        var maxTop = -Double.MAX_VALUE
        var maxBottom = Double.MAX_VALUE
        for (component in placedComponents) {
            val rightPadding = componentRuntime.rightPadding.coerceAtLeast(component.second[4])
            val leftPadding = componentRuntime.leftPadding.coerceAtLeast(component.second[5])

            val intersectingX = componentRuntime.x + componentRuntime.width / 2 + rightPadding > component.second[0] - component.second[2] / 2 + 1 && componentRuntime.x - componentRuntime.width / 2 + 1 < component.second[0] + component.second[2] / 2 + leftPadding

            val topPadding = componentRuntime.bottomPadding.coerceAtLeast(component.second[6])
            val canInteractTop = component.second[1] < componentRuntime.y
            if (intersectingX && canInteractTop) {
                val componentBottom = component.second[1] + component.second[3] / 2 + topPadding
                maxTop = maxTop.coerceAtLeast(componentBottom)
            }

            val bottomPadding = componentRuntime.topPadding.coerceAtLeast(component.second[7])
            val canInteractBottom = component.second[1] > componentRuntime.y
            if (intersectingX && canInteractBottom) {
                val componentTop = component.second[1] - component.second[3] / 2 - bottomPadding
                maxBottom = maxBottom.coerceAtMost(componentTop)
            }
        }
        return 0.0.coerceAtLeast(maxBottom - maxTop)
    }

    override fun getPaddingValue(componentRuntime: Component.Runtime): Double {
        return 0.0
    }

    override fun getColorValue(componentRuntime: Component.Runtime): Color {
        return null!!
    }

    override fun getStringValue(componentRuntime: Component.Runtime?): String {
        return null!!
    }

    override fun getCornerRadiusValue(componentRuntime: Component.Runtime): Double {
        return 0.0
    }
}